cura/UI/WelcomePagesModel.py

```diff
@@ -0,0 +1,220 @@
+# Copyright (c) 2019 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+from collections import deque
+import os
+from typing import TYPE_CHECKING, Optional, List, Dict, Any, Deque
+
+from PyQt5.QtCore import QUrl, Qt, pyqtSlot, pyqtProperty, pyqtSignal
+
+from UM.Logger import Logger
+from UM.Qt.ListModel import ListModel
+from UM.Resources import Resources
+
+if TYPE_CHECKING:
+    from PyQt5.QtCore import QObject
+
+#
+# This is the Qt ListModel that contains all welcome pages data. Each page is a page that can be shown as a step in the
+# welcome wizard dialog. Each item in this ListModel represents a page, which contains the following fields:
+#
+#  - id           : A unique page_id which can be used in function goToPage(page_id)
+#  - page_url     : The QUrl to the QML file that contains the content of this page
+#  - next_page_id : (OPTIONAL) The next page ID to go to when this page finished. This is optional. If this is not
+#                   provided, it will go to the page with the current index + 1
+#  - should_show_function : (OPTIONAL) An optional function that returns True/False indicating if this page should be
+#                           shown. By default all pages should be shown. If a function returns False, that page will
+#                           be skipped and its next page will be shown.
+#
+# Note that in any case, a page that has its "should_show_function" == False will ALWAYS be skipped.
+#
+class WelcomePagesModel(ListModel):
+
+    IdRole = Qt.UserRole + 1  # Page ID
+    PageUrlRole = Qt.UserRole + 2  # URL to the page's QML file
+    NextPageIdRole = Qt.UserRole + 3  # The next page ID it should go to
+
+    def __init__(self, parent: Optional["QObject"] = None) -> None:
+        super().__init__(parent)
+
+        self.addRoleName(self.IdRole, "id")
+        self.addRoleName(self.PageUrlRole, "page_url")
+        self.addRoleName(self.NextPageIdRole, "next_page_id")
+
+        self._pages = []  # type: List[Dict[str, Any]]
+
+        self._current_page_index = 0
+        # Store all the previous page indices so it can go back.
+        self._previous_page_indices_stack = deque()  # type: Deque[int]
+
+    allFinished = pyqtSignal()  # emitted when all steps have been finished
+    currentPageIndexChanged = pyqtSignal()
+
+    @pyqtProperty(int, notify = currentPageIndexChanged)
+    def currentPageIndex(self) -> int:
+        return self._current_page_index
+
+    # Returns a float number in [0, 1] which indicates the current progress.
+    @pyqtProperty(float, notify = currentPageIndexChanged)
+    def currentProgress(self) -> float:
+        return self._current_page_index / len(self._items)
```

Comment:
This can cause a ZeroDivisionError (len can be 0)

Reply:
Fixed in 3ad79a588